### PR TITLE
Modest memory savings in date_histogram>terms

### PR DESCRIPTION
**Pull request by nik9000**
This saves 16 bytes of memory per bucket for some aggregations.
Specifically, it kicks in when there is a parent bucket and we have a good
estimate on its upper bound cardinality, and we have good estimate on the
per-bucket cardinality of this aggregation, and both those upper bounds
will fit into a single `long`.

That sounds unlikely, but there is a fairly common case where we have it:
a `date_histogram` followed by a `terms` aggregation powered by global
ordinals. This is common enough that we already had at least two rally
operations for it:
* `date-histo-string-terms-via-global-ords`
* `filtered-date-histo-string-terms-via-global-ords`

Running those rally tracks shows that the space savings yields a small
but statistically significant perform bump. The 90th percentile service
time drops by about 4% in the unfiltered case and 1% for the filtered
case. That's not great but it good to know saving 16 bytes doesn't slow
us down.

```
|       50th percentile latency |          date-histo | 3185.77 | 3028.65 | -157.118 | ms |
|       90th percentile latency |          date-histo | 3237.07 | 3101.32 | -135.752 | ms |
|      100th percentile latency |          date-histo | 3270.53 |  3178.7 | -91.8319 | ms |
|  50th percentile service time |          date-histo | 3181.55 | 3024.32 | -157.238 | ms |
|  90th percentile service time |          date-histo | 3232.91 | 3097.67 | -135.238 | ms |
| 100th percentile service time |          date-histo | 3266.63 | 3175.08 | -91.5494 | ms |
|       50th percentile latency | filtered-date-histo | 1349.22 | 1331.94 | -17.2717 | ms |
|       90th percentile latency | filtered-date-histo | 1402.71 |  1383.7 | -19.0131 | ms |
|      100th percentile latency | filtered-date-histo | 1412.41 |  1397.7 | -14.7139 | ms |
|  50th percentile service time | filtered-date-histo | 1345.18 |  1326.2 | -18.9806 | ms |
|  90th percentile service time | filtered-date-histo | 1397.24 | 1378.14 | -19.1031 | ms |
| 100th percentile service time | filtered-date-histo | 1406.69 | 1391.63 | -15.0529 | ms |
```

The microbenchmarks for `LongKeyedBucketOrds`, the interface we're
targeting, show a performance boost on the method in the path of about
13%. This is obvious not the entire hot path, given that th 13% savings
translated to a 4% performance savings over the whole agg. But its
something.

```
Benchmark                                            Mode  Cnt  Score   Error  Units
multiBucketMany                                      avgt   10  9.956 ± 0.133  ns/op
multiBucketManySmall                                 avgt   10  8.739 ± 0.014  ns/op
singleBucketIntoMulti                                avgt   10  7.681 ± 0.031  ns/op
singleBucketIntoMultiSmall                           avgt   10  7.003 ± 0.451  ns/op
singleBucketIntoSingleImmutableBimorphicInvocation   avgt   10  6.176 ± 0.031  ns/op
singleBucketIntoSingleImmutableMonmorphicInvocation  avgt   10  6.534 ± 0.017  ns/op
singleBucketIntoSingleMutableBimorphicInvocation     avgt   10  7.727 ± 0.009  ns/op
singleBucketIntoSingleMutableMonmorphicInvocation    avgt   10  7.438 ± 0.037  ns/op
```

While I was touching the JMH benchmarks for `LongKeyedBucketOrds` I took
the opportunity to try and make the runs that collect from a single
bucket more comparable to the ones that collect from many buckets. It
only seemed fair.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/LongKeyedBucketOrds.java [hunk 1]
```diff
@@ -274,4 +311,129 @@ public void close() {
...
+        private final long owningBucketOrdMask;
+
+        public FromManySmall(BigArrays bigArrays, int owningBucketOrdShift) {
+            ords = new LongHash(2, bigArrays);
+            this.owningBucketOrdShift = owningBucketOrdShift;
+            this.owningBucketOrdMask = -1L << owningBucketOrdShift;
+        }
+
+        private long encode(long owningBucketOrd, long value) {
+            // This is in the critical path for collecting lots of aggs. Be careful of performance.
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/LongKeyedBucketOrds.java:327]: This isn't 100% true - it's only the hot path for a `terms` agg under an agg that knows its cardinality like `range` or `filters`. Or a `date_histogram` when it's been rewritten into `range` or `filters`.

**Review by nik9000 — COMMENTED** [review]
Not bad for an evening of bit twiddling!

**Comment by nik9000** [comment]
> Benchmark                                            Mode  Cnt   Score   Error  Units
> multiBucketMany                                      avgt    5  10.038 ± 0.009  ns/op
> multiBucketManySmall                                 avgt    5   8.738 ± 0.029  ns/op
> singleBucketIntoMulti                                avgt    5   7.701 ± 0.073  ns/op
> singleBucketIntoSingleImmutableBimorphicInvocation   avgt    5   6.160 ± 0.029  ns/op
> singleBucketIntoSingleImmutableMonmorphicInvocation  avgt    5   6.571 ± 0.043  ns/op
> singleBucketIntoSingleMutableBimorphicInvocation     avgt    5   7.714 ± 0.010  ns/op
> singleBucketIntoSingleMutableMonmorphicInvocation    avgt    5   7.459 ± 0.017  ns/op

I'm rerunning these to pick up some last minute changes I've made. It takes 45 minutes though.

**Comment by nik9000** [comment]
I've updated the JMH.

**Review by not-napoleon — APPROVED** [review]
I left a bunch of notes on places comments could be cleaned up a bit, but overall this looks good to me and in line with what we discussed a few weeks ago.  Nice bit of cleverness.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/LongKeyedBucketOrds.java (hunk 1 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/LongKeyedBucketOrds.java:329]: Yeah. I guess the untrue bit is "lots". I'll just remove that word.

File: benchmarks/src/main/java/org/elasticsearch/benchmark/search/aggregations/bucket/terms/LongKeyedBucketOrdsBenchmark.java [hunk 2]
```diff
@@ -134,28 +151,69 @@ public void singleBucketIntoSingleMutableBimorphicInvocation(Blackhole bh) {
...
             bh.consume(ords);
         }
     }
 
     /**
-     * Emulates an aggregation that collects from many buckets.
+     * Emulates an aggregation that collects from a single bucket "by accident".
+     * This can happen if an aggregation is under, say, a {@code terms}
+     * aggregation and there is only a single value for that term in the index
+     * but we can't tell that up front.
```
[the remark this answers is not in the document]
Comment (nik9000) [reply benchmarks/src/main/java/org/elasticsearch/benchmark/search/aggregations/bucket/terms/LongKeyedBucketOrdsBenchmark.java:174]: :+1:

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/LongKeyedBucketOrds.java [hunk 3]
```diff
@@ -14,17 +14,41 @@
...
      */
     public static LongKeyedBucketOrds build(BigArrays bigArrays, CardinalityUpperBound cardinality) {
         return cardinality.map(estimate -> estimate < 2 ? new FromSingle(bigArrays) : new FromMany(bigArrays));
     }
 
+    /**
+     * Build a {@link LongKeyedBucketOrds} who's values have known bounds.
+     */
+    public static LongKeyedBucketOrds buildForValueRange(BigArrays bigArrays, CardinalityUpperBound cardinality, long min, long max) {
+        return cardinality.map(cardinalityUpperBound -> {
```
[the remark this answers is not in the document]
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/LongKeyedBucketOrds.java:34]: I can make the change, sure! It is an int which is a little surprising, I think. So it's useful to have.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/LongKeyedBucketOrds.java [hunk 4]
```diff
@@ -274,4 +313,135 @@ public void close() {
...
+            }
+            if ((value & owningBucketOrdMask) != 0) {
+                return -1;
+            }
+            return ords.find(encode(owningBucketOrd, value));
+        }
+
+        @Override
+        public long get(long ordinal) {
+            return ords.get(ordinal) & ~owningBucketOrdMask;
```
[the remark this answers is not in the document]
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/LongKeyedBucketOrds.java:365]: Its getting the value associated with an ordinal its returned from a call to `add`. It takes an `ordinal`, not an `owningBucketOrdinal`. That doesn't sound sensible when I type it though. It's true, and it makes sense. I promise you. But it doesn't sound like it makes sense.

**Comment by nik9000** [comment]
run elasticsearch-ci/bwc

**Comment by nik9000** [comment]
@elasticmachine, test this please.

**Pull request merged**